// react-relay/lib/index.js
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 * @format
 */
// flowlint ambiguous-object-type:error
'use strict';

var ReactRelayContext = require('./ReactRelayContext');

var ReactRelayFragmentContainer = require('./ReactRelayFragmentContainer');

var ReactRelayLocalQueryRenderer = require('./ReactRelayLocalQueryRenderer');

var ReactRelayPaginationContainer = require('./ReactRelayPaginationContainer');

var ReactRelayQueryRenderer = require('./ReactRelayQueryRenderer');

var ReactRelayRefetchContainer = require('./ReactRelayRefetchContainer');

var RelayRuntime = require('relay-runtime');

/**
 * The public interface to React Relay.
 */
module.exports = {
  ConnectionHandler: RelayRuntime.ConnectionHandler,
  QueryRenderer: ReactRelayQueryRenderer,
  LocalQueryRenderer: ReactRelayLocalQueryRenderer,
  MutationTypes: RelayRuntime.MutationTypes,
  RangeOperations: RelayRuntime.RangeOperations,
  ReactRelayContext: ReactRelayContext,
  applyOptimisticMutation: RelayRuntime.applyOptimisticMutation,
  commitLocalUpdate: RelayRuntime.commitLocalUpdate,
  commitMutation: RelayRuntime.commitMutation,
  createFragmentContainer: ReactRelayFragmentContainer.createContainer,
  createPaginationContainer: ReactRelayPaginationContainer.createContainer,
  createRefetchContainer: ReactRelayRefetchContainer.createContainer,
  fetchQuery: RelayRuntime.fetchQuery,
  graphql: RelayRuntime.graphql,
  readInlineData: RelayRuntime.readInlineData,
  requestSubscription: RelayRuntime.requestSubscription
};
